### ISSUE TITLE
tracking provenance of utility code

### ISSUE THREAD

**Issue by pabigot** [zephyrproject-rtos/zephyr#26197]
tl;dr: `sys/util.h` has non-trivial code that was adapted from a Stack Overflow answer but lacks clear documentation of copyright and licensing.

> IANAL but this post was edited in 2017, which is after [this policy](https://stackoverflow.com/questions/12447557/can-we-have-recursive-macros/12540675#12540675) went into effect, so we appear to be compliant because of [this comment](https://github.com/zephyrproject-rtos/zephyr/pull/26134/files#diff-418e933b5df2b11420aa260d71e1fd5dR91), though a license expert should confirm if that continues to be a concern. Of course, this PR isn't adding that code, just moving it around.

The links above don't make sense to me: the first is to the same stackoverflow answer that's referenced in the header, not a (Zephyr?) policy; and the second to the entire set of changes in this PR, not a comment.

Regardless, I'm not saying there's something to be done in this PR, just that the provenance of the code in that header and its licensing is not entirely clear.  The copyright assignment to Wind River Systems 2011-2014 as indicated in current master is pretty obviously wrong, though.  At least some of this was originally added in 0a4808b75ca2b8a8836eccd4d5a2069abebd6181 via #4692.  https://stackoverflow.com/help/licensing is also relevant to the question (but not this PR); the `SPDX-License-Identifier` perhaps should be `(Apache-2.0 AND CC-BY-SA-3.0)`.

I'll just make this comment an issue in its own right so we can think about it while we carry on with more immediate concerns.

_Originally posted by @pabigot in https://github.com/zephyrproject-rtos/zephyr/pull/26134#issuecomment-644269406_

**Comment by pabigot** [comment]
https://github.com/zephyrproject-rtos/zephyr/pull/26134#issuecomment-644306009 provides the correct link https://meta.stackexchange.com/questions/271080/the-mit-license-clarity-on-using-code-on-stack-overflow-and-stack-exchange so this file should probably be `(Apache-2.0 AND MIT)`.

The more general point is that we should be cognizant of cases where code is being adopted from other sources and ensure license compliance and provenance are included in PRs.
